Manageiq group module

evertmulder opened the pull request:
##### SUMMARY
Added a module to manage groups in ManageIQ

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
manageiq_group

##### ANSIBLE VERSION
```
 ansible --version
ansible 2.6.1
  config file = None
  configured module search path = [u'/Users/emulder/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Mar 21 2018, 16:34:57) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

##### ADDITIONAL INFORMATION
New module for creating groups in manageiq

Commit:
ManageIQ module to manage groups

yaacov commented:
shipit

@cben PTAL

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,596 @@
+  role:
+    description:
+    - The the group role name
+    - The C(role_id) has president over the C(role) when supplied.
```
has precedence?

evertmulder replied:
😊 typo

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,596 @@
+    required: false
+    default: null
+  managed_filters:
+    description: A list of list of tag categories
```
Can you add some more detail on what the structure means?  How is [[a, b, c, d]] different from [[a, b], [c, d]]?

evertmulder replied:
The filters in the manageiq API expect a list of tags per category, for example adding tags in the category prov_max_cpu and department would look something like this:
```yaml  
managed_filters:
    - - "/managed/prov_max_cpu/1"
      - "/managed/prov_max_cpu/3"
      - "/managed/prov_max_cpu/4"
    - - "/managed/department/defense"
      - "/managed/department/engineering"
```

This is how I currently implemented the `managed_filters`. It matches the ManageIQ API one on one. But I agree this is not that intuitive. ManageIQ tags always seem to start with '/managed', I don't know why...

Various options that crossed my mind, some where: 

```yaml
managed_filters:
  - "prov_max_cpu":
    - "3"
    - "4"
    state: "present"
  - "department":
    - "defense"
    - "engineering"
    state: "absent"
```

Or specifying tags individually 

```yaml
managed_filters:
  - prov_max_cpu : "3"
    state: "present"
  - prov_max_cpu : "4"
    state: "present"
  - prov_max_cpu : "5"
    state: "present"
  - department : "defense"
    state: "present"
  - department : "engineering"
    state: "absent"   
```

In my opinion I would like to describing the desired state per category would work best.

cben replied:
Thanks.  This still doesn't explain nor do API docs or UI docs what's the meaning of these filters :-(.  Opened https://github.com/ManageIQ/manageiq_docs/issues/977.

The latter options do seem friendlier.
Would `state` also allow getting rid of `managed_filters_merge_mode`?
Conversely, perhaps a separate `managed_filters_merge_mode` allows dropping `state` and simplifying even further:
```
managed_filters:
  "prov_max_cpu": ["3", "4"]
  "department": []
managed_filters_merge_mode: replace
```
BTW, there are 3 behaviors that could make sense: full replacement of `filters.managed`, replacing each category, and merging individual tags.

How important is it for this module to support partial updates of filters?
Personally, I'd start with full replace only, with cleanest possible structure.

Anyway, LGTM with any structure, just please document it a bit clearer in description.

cben replied:
Actually with the example you updated, this text is good enough. However we explain it, it won't be clear until one looks at examples...  Perhaps slight tweak "A list of list of tags, grouped by category".

evertmulder replied:
I will use your suggestion and update the description.

evertmulder replied:
I liked you suggestion

```yml
managed_filters:
  "prov_max_cpu": ["3", "4"]
  "department": []
managed_filters_merge_mode: replace
```
And only support managed_filters_merge_mode: replace and merge (not merging of separate tags)

Much betten now. Thanks.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,596 @@
+    created_on:
+      description: Group creation date
+      returned: success
+      type: int
```
`2018-08-12T08:37:55+00:00` is not an int :)

evertmulder replied:
changed to string

Commit:
Fixed typos, more examples on managed_filters.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+            new_filters_resource['belongsto'] = current_filters['belongsto']
+        else:
+            if belongsto_filters_merge_mode == 'merge':
+                current_belongsto_set.update(new_belongsto_set)
```
Is this deliberately different from `merge_values_nested_array(current_managed, managed_filters)` below?
Ouch, I see, belongsto is 1 level while managed is 2 levels:
```
    "filters" : {
      "belongsto" : [ "/managed/area/1", "/managed/area/2", "/managed/area/3" ],
      "managed" : [[ "/managed/infra/1", "/managed/infra/2"], ["/managed/other/3"]]
    }
```
LGTM :+1:

evertmulder replied:
Yep the belongs to is easy to merge...

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+            description=group['description'],
+            role_name=role_name,
+            tenant_name=group['tenant']['name'],
+            filters=filters,
```
I think it's nice for ansible returned structure to be similar to module arguments where possible.
Can you expand this into `managed_filters` and `belongsto_filters`?

evertmulder replied:
Yes, I will change this.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+    def tenant(self, tenant_id, tenant_name):
+        """ Search for tenant entity by name or id
+        Returns:
+            the tenant entity, None no id or name was supplied
```
None **if** no id ...

evertmulder replied:
Good catch. Fixed

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+        if current_values and (not updated_values):
+            return False
+        if (not current_values) and updated_values:
+            return False
```
Are these special cases (and some in the `merge_`) function intended to support None?
IMHO it would simplify code to normalize them to `[]` or `set()` in caller.

evertmulder replied:
I tried to create a set from the nested array, but this is not allowed.

evertmulder replied:
Did a rewrite of this part. Now noralized the miq values and the user input to a dict of string arrays.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+            id=group['id'],
+            description=group['description'],
+            role_name=role_name,
+            tenant_name=group['tenant']['name'],
```
consider calling these `role`, `tenant` with _name suffix, to match module arguments.

evertmulder replied:
Yes, probably cleaner.
Fixed

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+        # organize the values per category name
+        for sub in current_values:
+            sub.sort()
+            key = sub[0].split('/')[2]
```
just in case there's ever something not starting `/managed/`, safer to use `[0:2]` here.

Also, consider extracting this list-of-lists to dict-of-sorted-lists transform to a separate method.
After transforming desired & current, `have_same_values_nested_array` would become simple ==,
merging would be simpler, and transform back is simple `.values()`.

evertmulder replied:
Refactored this.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+
+        # add and update current categories
+        for key in new_values_categories:
+            current_values_categories[key] = new_values_categories[key]
```
This is a shallow merge, right?
That is, when both current and new contain same category, the whole category is replaced?
BTW, could write this simpler as `current_values_categories.update(new_values_categories)`.

This is not self-evident, could also be a deep merge (taking union of current | new tags in each category).
Can you clarify the managed_filters_merge_mode documentation?

evertmulder replied:
Added documentation in the code

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_group.py:
```diff
@@ -0,0 +1,601 @@
+        res = []
+
+        for key in current_values_categories:
+            res.append(current_values_categories[key])
```
I think this can be simplified to `return current_values_categories.values()`.

evertmulder replied:
Yes. for python3 is must be return list(current_values_categories.values()), still cleaner.

cben reviewed (commented):
Sorry for delay, back from holidays.
Small nits and suggestions (many subjective, feel free to push back), overall great :+1:

Commit:
Simplified the input if managed_filters
Change return values role_name and tenant_name to role and tenant to macht input params

Commit:
Return managed_filters and belongsto_filters instead of filters to better mach the input params

evertmulder commented:
Hi @cben, 
thanks for your great feedback. Changed a bunch of stuf based on your comments. Could you take another look?

Thanks.

cben commented:
Oops, forgot all about this.  Re-read superficially but definitely looks nice :sparkles:
shipit

cben commented:
bot_status

cben commented:
@abellotti @gtanzillo @yaacov @zgalor this needs more shipits... (And then maybe ansible core review, i'm never sure...)

yaacov commented:
shipit

gundalow commented:
@evertmulder Thank you for this PR
@cben @yaacov Thank you for the reviews

Merged into `devel` for release in Ansible 2.8

@cben Once the PR gets the `shipit` label we will merge it
